File: Graphs/LC_463.cpp
class Solution {
public:
    int dfs(vector<vector<int>>& grid, int row, int col, int m, int n, unordered_map<int, unordered_map<int, int>>& seen){
        if (row < 0 || row == m || col < 0 || col == n || grid[row][col] == 0){
            return 1;
        }
        if (seen.count(row) && seen[row].count(col)){
            return 0;
        }
        seen[row][col] = 1;
        int t = 0, b = 0, l = 0, r = 0;
        t = dfs(grid, row-1, col, m, n, seen);
        b = dfs(grid, row+1, col, m, n, seen);
        l = dfs(grid, row, col-1, m, n, seen);
        r = dfs(grid, row, col+1, m, n, seen);
        return t + b + l + r;
    }
    int islandPerimeter(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        int ans = 0;
        unordered_map<int, unordered_map<int, int>> seen;
        for (int row = 0; row < m; row++){
            for (int col = 0; col < n; col++){
                if (grid[row][col] == 1){
                    ans = dfs(grid, row, col, m, n, seen);
                    return ans;
                }
            }
        }
        return ans;
    }
};

File: Graphs/LC_463.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        visited = set()
        def dfs(grid, row, col, visited):
            # border cases
            if row < 0 or row == m or col < 0 or col == n or grid[row][col] == 0:
                return 1
            if (row, col) in visited:
                return 0
            visited.add((row, col))
            ans = dfs(grid, row+1, col, visited)
            ans += dfs(grid, row-1, col, visited)
            ans += dfs(grid, row, col+1, visited)
            ans += dfs(grid, row, col-1, visited)
            return ans
        for row in range(m):
            for col in range(n):
                if grid[row][col] == 1:
                    return dfs(grid, row, col, visited)